docs/design-documents/20230228-connector-lifecycle-methods.md

```diff
@@ -0,0 +1,238 @@
+# Connector Lifecycle Methods
+
+## Introduction
+
+Connector Lifecycle Methods are additional integration points that would let
+connectors execute certain operations when the connector is first created, when
+its configuration is updated or when it gets deleted.
+
+### Background
+
+We first saw a need for this in the Postgres connector, which can create and
+manage its own logical replication slot. We run into a cleanup problem - if the
+logical replication slot is created by the connector it must also be deleted by
+the connector, so the connector currently deletes it in `Teardown`. This defeats
+the whole purpose of a logical replication slot because the connector loses all
+state. Next time it starts it will create a new logical replication slot and
+miss all changes that happened while the pipeline was not running.
+
+Introducing lifecycle methods would allow the connector to create the logical
+replication slot the first time the pipeline runs and remove it once the
+connector gets deleted. Or more generally - lifecycle methods would allow a
+connector to execute initializations on creation and clean up after itself on
+deletion. Additionally, we need to take into account that the configuration can
+change between runs and we need to notify the connector about these changes.
+
+### Goals
+
+- Lifecycle methods should be optional to implement.
+- The introduction of lifecycle methods must be backwards compatible with
+  existing connectors.
+- The connector needs to be notified when it was first created*.
+- The connector needs to be notified when its configuration has changed*.
+- The connector needs to be notified when it gets deleted.
+
+\*Actually when the pipeline is first started after that happened. See
+[implementation options](#implementation-options) for more details.
```

Comment:
I'm not following this line - is this about the order of operations (pipeline starting after the first lifecycle event for creation)?

Reply:
I see I created just confusion with this line, I'm inclined to delete it 😅 Read on, the answers are lower in the document.

Reply:
yeah, the rest of the document addressed all my questions. I'm good with deleting this line, it did throw me off a bit 😄